datafusion/core/src/datasource/file_format/parquet.rs

```diff
@@ -543,6 +574,172 @@ async fn fetch_statistics(
     Ok(statistics)
 }
 
+/// Implements [`DataSink`] for writing to a parquet file.
+struct ParquetSink {
+    /// Config options for writing data
+    config: FileSinkConfig,
+}
+
+impl Debug for ParquetSink {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ParquetSink").finish()
+    }
+}
+
+impl DisplayAs for ParquetSink {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                write!(
+                    f,
+                    "ParquetSink(writer_mode={:?}, file_groups=",
+                    self.config.writer_mode
+                )?;
+                FileGroupDisplay(&self.config.file_groups).fmt_as(t, f)?;
+                write!(f, ")")
+            }
+        }
+    }
+}
+
+impl ParquetSink {
+    fn new(config: FileSinkConfig) -> Self {
+        Self { config }
+    }
+
+    /// Builds a parquet WriterProperties struct, setting options as appropriate from TaskContext options
+    fn parquet_writer_props_from_context(
+        &self,
+        context: &Arc<TaskContext>,
+    ) -> WriterProperties {
+        let parquet_context = &context.session_config().options().execution.parquet;
+        let mut builder = WriterProperties::builder()
+            .set_created_by(parquet_context.created_by.clone())
+            .set_data_page_row_count_limit(parquet_context.data_page_row_count_limit)
+            .set_data_page_size_limit(parquet_context.data_pagesize_limit);
```

Comment:
it feels like there are some other properties to copy across to (like created by, etc)

Reply:
Yes more to do here. Some of the WriterProperties configs are not primitive types (such as compression and statistics). Still working through this, but the plan is to have the Session level config be a string and implement TryFrom(String) for these types. Of course since these types/traits are defined in other crates, so we would either need to implement TryFrom in arrow-rs or have a wrapper type in Datafusion.

Let me know if we already have a pattern to handle configs that ultimately need to map to non primitive types that I just missed. 🤔

Reply:
In this setup, it would also be better UX to push parsing the string up to when the user initially sets the config, so the user can get immediate feedback rather than a runtime error if they provide an invalid option. E.g.:

```set execution.parquet.compression = exotic_compression_123;
PropertyError: unknown or unsupported parquet compression codec: exotic_compression_123
```

Otherwise, it would seem that the setting is valid and only fail much later when a write is attempted.

Reply:
Agreed on not doing too much in this one PR. In the current state of this PR we support customizing default parquet writer properties via session config options. Column-by-column customization support needs to be added, and table and/or statement level override support needs to be added. Config validation framework + moving some of the validation code from this PR up to when configs are parsed would be a nice enhancement as well. 

We could cut issues to track for future PRs though so this one does not get too complicated or take too long.